pallets/cf-reward/src/lib.rs

```diff
@@ -0,0 +1,208 @@
+        NotEnoughFunds,
+        /// Initalization before the last lease period is over
+        OngoingLease,
+    }
+}
+
+
+decl_module! {
+    pub struct Module<T: Config> for enum Call where origin: <T as frame_system::Config>::Origin {
+        // Activate errors
+        type Error = Error<T>;
+
+        // Activate events
+        fn deposit_event() = default;
+
+        /// Initialize the module.
+        /// This dispatchable function must be called from a root account, so that other accounts
+        /// can not block the module via a random initialization. Furthermore it is important, that
+        /// the module has enough funds, as a later funding is not possible.
+        // TODO: Add correct weight function
+        #[weight = 10_000]
+        fn initialize(origin, source: T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
+            ensure_root(origin)?;
+
+            let now = <frame_system::Module<T>>::block_number();
+            ensure!(now > <NextInit<T>>::get(), Error::<T>::OngoingLease);
+
+            let target = &MODULE_ID.into_account();
+            // TODO: Should we choose: KeepAlive or AllowDeath here?
+            T::Currency::transfer(
+                &source,
+                &target,
+                amount,
+                AllowDeath,
+            )?;
+
+            <NextInit<T>>::put(now + T::LeasePeriod::get());
+
+            Self::deposit_event(RawEvent::RewardModuleInitalized);
+
+            Ok(())
+        }
+    }
+}
+
+impl<T: Config> Module<T> {
+    fn convert_to_native(contribution: T::RelayChainBalance) -> <T::Currency as Currency<T::AccountId>>::Balance {
+        contribution.into() * <ConversionRate<T>>::get()
+    }
+}
+
+impl<T: Config> Reward for Module<T> {
+    type ParachainAccountId = T::AccountId;
+    type ContributionAmount = T::RelayChainBalance;
+
+    fn reward(who: Self::ParachainAccountId, contribution: Self::ContributionAmount) -> DispatchResult {
+        let schedule = pallet_vesting::VestingInfo {
+            locked: Self::convert_to_native(contribution),
+            per_block: <<T as pallet_vesting::Config>::BlockNumberToBalance>::convert(T::VestingPeriod::get()),
+            starting_block: <frame_system::Module<T>>::block_number(),
+        };
+
+        let from: <T as frame_system::Config>::AccountId = MODULE_ID.into_account();
+        let target = <T::Lookup as StaticLookup>::unlookup(who.clone());
+
+        <pallet_vesting::Module<T>>::vested_transfer(
+            T::Origin::from(RawOrigin::Signed(from)),
+            target,
+            schedule
+        ).map_err(|vest_err| {
+            match vest_err {
+                DispatchError::Module { index: _, error: _, message } => {
+                    message.map_or(vest_err, |msg| {
+                        match msg {
+                            "InsufficientBalance" => Into::<DispatchError>::into(Error::<T>::NotEnoughFunds),
+                            "AmountLow" => { vest_err
```

Comment:
Amount low is the minimum amount target gets, right? If we want to enforce the minimum amount, shouldn't we do the check before we reach here?

Reply:
No, its the minimum amount that is necessary to create a vesting schedule. Currently this [1000*CFG](https://github.com/centrifuge/centrifuge-chain/blob/20dd92afbb64ef062543f0f8767dfe120ad6f96c/runtime/src/lib.rs#L841)

Reply:
I updated the current state. I first thought that we will be able to transfer any additional funds to the module. But with the module account being stored in the module's storage, this should be straightforward.

Reply:
I'm confused between `InsufficientBalance` and `AmountLow` now. Doesn't both represent same now ?

Reply:
The first one refers to an error triggered by the balances pallet, and the second one to the vesting pallet. I am translating the `InsufficientBalance` error to a reward pallet specific one `NotEnoughFunds`. Thought this gives us more information.